walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,80 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return possibleTimes;
+    }
+
+    possibleTimes.add(TimeRange.WHOLE_DAY);
+
+    if (request.getAttendees().size() == 0) {
+        return possibleTimes;
+    }
+
+    // Only update times for events that have attendees from the meeting request
+    for (Event event : events) {
+        for (String attendee : request.getAttendees()) {
+            if (event.getAttendees().contains(attendee)) {
+                updatePossibleTimes(possibleTimes, event.getWhen(), request.getDuration());
+            }
+        }
+    }
+
+    return possibleTimes;
+  }
+
+  /** 
+   * Update possible meeting times by excluding time ranges of an event
+   */
+  private void updatePossibleTimes(Collection<TimeRange> possibleTimes, TimeRange eventTime, long requestDuration) {
+    Collection<TimeRange> rangesToModify = new ArrayList<TimeRange>();
+    for (TimeRange timeSlot : possibleTimes) {
+        if (timeSlot.contains(eventTime.start()) || timeSlot.contains(eventTime.end())) {
+            rangesToModify.add(timeSlot);
+        }
+    }
+
+    splitTimeRanges(possibleTimes, rangesToModify, eventTime);
+    removeInsufficientSlots(possibleTimes, requestDuration);
+
+  }
+
+  /** 
+   * For ranges that overlap the event, remove the invalid parts from the range
+   */
+  private void splitTimeRanges(Collection<TimeRange> possibleTimes, Collection<TimeRange> rangesToModify, TimeRange eventTime) {
+    for (TimeRange timeSlot : rangesToModify) {
+        if (timeSlot.contains(eventTime.start())) {
+            possibleTimes.add(TimeRange.fromStartEnd(timeSlot.start(), eventTime.start(), false));
+        }
+
+        if (timeSlot.contains(eventTime.end())) {
+            possibleTimes.add(TimeRange.fromStartEnd(eventTime.end(), timeSlot.end(), false));
+        }
+
+        possibleTimes.remove(timeSlot);
+    }
+  }
+
+  /** 
+   * Remove possible meeting times that are not long enough for the requested meeting
+   */
+  private void removeInsufficientSlots(Collection<TimeRange> possibleTimes, long requestDuration) {
+      Collection<TimeRange> rangesToRemove = new ArrayList<TimeRange>();
+      for (TimeRange range : possibleTimes) {
+          if (range.duration() < requestDuration) {
+              rangesToRemove.add(range);
+          }
+      }
+
+      possibleTimes.removeAll(rangesToRemove);
   }
```

Comment:
Its a bit hard to follow whats going on here. Could you explain?

Reply:
I tried to explain with the new comments - are there any functions that are confusing in particular?